app/models/product.rb

```diff
@@ -14,7 +14,13 @@ class Product < ApplicationRecord
     validates :price, numericality: { greater_than_or_equal_to: 0.01 }
     validates :discount_price, numericality: {less_than_or_equal_to: :price, message: "Discount Price can't be greater than Original Price"}, if: :price 
   end
-
+  
+  before_validation :set_discount_price
+  before_destroy :ensure_not_referenced_by_any_line_item
```

Comment:
If this method is called as a callback, should this really add errors to `errors` object